data/helpers.d/system

```diff
@@ -56,6 +56,100 @@ ynh_get_debian_release () {
 	echo $(lsb_release --codename --short)
 }
 
+# Start (or other actions) a service,  print a log in case of failure and optionnaly wait until the service is completely started
+#
+# usage: ynh_systemd_action [-n service_name] [-a action] [ [-l "line to match"] [-p log_path] [-t timeout] [-e length] ]
+# | arg: -n, --service_name= - Name of the service to reload. Default : $app
+# | arg: -a, --action=       - Action to perform with systemctl. Default: start
+# | arg: -l, --line_match=   - Line to match - The line to find in the log to attest the service have finished to boot.
+#                              If not defined it don't wait until the service is completely started.
+#                              WARNING: When using --line_match, you should always add `ynh_clean_check_starting` into your
+#                                `ynh_clean_setup` at the beginning of the script. Otherwise, tail will not stop in case of failure
+#                                of the script. The script will then hang forever.
+# | arg: -p, --log_path=     - Log file - Path to the log file. Default : /var/log/$app/$app.log
+# | arg: -t, --timeout=      - Timeout - The maximum time to wait before ending the watching. Default : 300 seconds.
+# | arg: -e, --length=       - Length of the error log : Default : 20
+ynh_systemd_action() {
+    # Declare an array to define the options of this helper.
+    declare -Ar args_array=( [n]=service_name= [a]=action= [l]=line_match= [p]=log_path= [t]=timeout= [e]=length= )
+    local service_name
+    local action
+    local line_match
+    local length
+    local log_path
+    local timeout
+
+    # Manage arguments with getopts
+    ynh_handle_getopts_args "$@"
+
+    local service_name="${service_name:-$app}"
+    local action=${action:-start}
+    local log_path="${log_path:-/var/log/$service_name/$service_name.log}"
+    local length=${length:-20}
+    local timeout=${timeout:-300}
+
+    # Start to read the log
+    if [[ -n "${line_match:-}" ]]
+    then
+        local templog="$(mktemp)"
+        # Following the starting of the app in its log
+        if [ "$log_path" == "systemd" ] ; then
+            # Read the systemd journal
+            journalctl --unit=$service_name --follow --since=-0 --quiet > "$templog" &
+            # Get the PID of the journalctl command
+            local pid_tail=$!
+        else
+            # Read the specified log file
+            tail -F -n0 "$log_path" > "$templog" &
+            # Get the PID of the tail command
+            local pid_tail=$!
+        fi
+    fi
+
+    echo "${action^} the service $service_name" >&2
+    systemctl $action $service_name \
```

Comment:
I'm wondering, why not using `yunohost service $action $service_name instead`? To deal with the lock?

Reply:
Which lock are you talking about ?

Reply:
I'm really sorry, I have to.

/!\ WARNING, PYTHON BELOW /!\

I'm talking about this: https://github.com/YunoHost/yunohost/blob/stretch-unstable/src/yunohost/service.py#L685

Reply:
My eyes are bleeding...

I think this lock is only for YunoHost services.

Reply:
I know that hotspot_ynh from labriqueinternet use a lock.

Reply:
I'm not really convinced when I'm asking that, so you can keep it like that!

Reply:
Hmm, let's forget about this --need_lock for now, it's indeed basically only for vpnclient / hotspot and is related to the use of `systemctl` with service themselves triggering `yunohost` commands ...

Will see in the future if we really need to adapt this helper. But there are other strategies to investigate like getting rid of using `yunohost` in the service.